Batteries/Data/BitVec.lean

```diff
@@ -0,0 +1,42 @@
+/-
+Copyright (c) 2024 François G. Dorais. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: François G. Dorais
+-/
+
+namespace BitVec
+
+theorem getElem_shifConcat (v : BitVec n) (b : Bool) (i) (h : i < n) :
+    (v.shiftConcat b)[i] = if i = 0 then b else v[i-1] := by
+  rw [← getLsbD_eq_getElem, getLsbD_shiftConcat, getLsbD_eq_getElem, decide_eq_true h,
+    Bool.true_and]
+
+@[simp] theorem getElem_shiftConcat_zero (v : BitVec n) (b : Bool) (h : 0 < n) :
+    (v.shiftConcat b)[0] = b := by simp [getElem_shifConcat]
+
+@[simp] theorem getElem_shiftConcat_succ (v : BitVec n) (b : Bool) (h : i + 1 < n) :
+    (v.shiftConcat b)[i+1] = v[i] := by simp [getElem_shifConcat]
+
+/-- `ofFnAux f` returns the `BitVec m` whose `i`th bit is `f i` when `i < m` -/
+@[inline] def ofFnAux (m : Nat) (f : Fin n → Bool) : BitVec m :=
+  Fin.foldr n (fun i v => v.shiftConcat (f i)) 0
+
+/-- `ofFn f` returns the `BitVec n` whose `i`th bit is `f i` -/
+abbrev ofFn (f : Fin n → Bool) : BitVec n := ofFnAux n f
```

Comment:
New functions creating BitVecs should have simp lemmas for `toNat`, `toFin`, and ideally also `toInt`.

Reply:
Doesn't that need the `Nat`/`Fin`/`Int` functions to already exist in some way? I would probably just implement those using the BitVec version.

Reply:
I added `Nat.ofBits` and `Fin.ofBits`, little endian only since that's the way `Nat.testBit` works. I will try to think about what the right `Int` version should be.

Reply:
Adding `Int.ofBits` would require upstreaming `Int.testBit` from Mathlib. Let me know if you think that's a good idea and I will get it done.

Reply:
I'm opting delaying the `Int` part from this PR. The `Nat` and `Fin` will shortly be moved to the prequel PR #1089